conjure-lib/src/main/java/com/palantir/conjure/java/lib/DoubleArrayList.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.lib;
+
+import java.util.AbstractList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.RandomAccess;
+
+public final class DoubleArrayList extends AbstractList<Double> implements RandomAccess {
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    private int size = 0;
+    private double[] elements;
+
+    public DoubleArrayList() {
+        this(16);
+    }
+
+    private void resizeIfNecessary(int toAdd) {
+        if (size + toAdd > elements.length) {
+            elements = Arrays.copyOf(elements, newCapacity(size + toAdd));
+        }
+    }
+
+    @SuppressWarnings("ThrowError")
+    private int newCapacity(int minCapacity) {
+        int oldCapacity = elements.length;
+        // increase by 50%
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity - minCapacity <= 0) {
+            if (elements.length == 0) {
+                return Math.max(10, minCapacity);
+            }
+            if (minCapacity < 0) {
+                throw new OutOfMemoryError();
+            }
+            return minCapacity;
+        }
+        return (newCapacity - MAX_ARRAY_SIZE <= 0) ? newCapacity : hugeCapacity(minCapacity);
+    }
+
+    @SuppressWarnings("ThrowError")
+    private static int hugeCapacity(int minCapacity) {
+        if (minCapacity < 0) {
+            throw new OutOfMemoryError();
+        }
+        return (minCapacity > MAX_ARRAY_SIZE) ? Integer.MAX_VALUE : MAX_ARRAY_SIZE;
+    }
+
+    public DoubleArrayList(int initialCapacity) {
+        elements = new double[initialCapacity];
```

Comment:
The Java `ArrayList` library uses a neat trick: they allocate a single static empty array so empty lists end up costing an object ptr & an integer.

Reply:
Cool! 86ef3c11

Reply:
Cool! 86ef3c11